### PR TITLE
Devel rtos timer2

### PR DESCRIPTION
**Pull request by scottwcpg**
A kernel timer driver using a 32KHz 32-bit down counter. Please review!!!!

### CODE REVIEW

File: drivers/timer/mchp_xec_rtos_timer.c [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+	return (RTMR_REGS->CTRL & MCHP_RTMR_CTRL_FW_HALT_EN);
+}
+
+static void timer_ien(void)
+{
+	GIRQ23_REGS->EN_SET = MCHP_RTMR_GIRQ_VAL;
+}
+
+/* SYSTICK elapsed function */
+#if 0
```
Comment (andyross) [inline drivers/timer/mchp_xec_rtos_timer.c]: Remove

File: drivers/timer/mchp_xec_rtos_timer.c [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+static volatile u32_t announced_cycles;
+
+static volatile u32_t elapsed_counts;
+
+static ALWAYS_INLINE void timer_clear_intr(void)
+{
+	GIRQ23_REGS->SRC = MCHP_RTMR_GIRQ_VAL;
+	GIRQ23_REGS->RESULT = GIRQ23_REGS->SRC;
+	NVIC_ClearPendingIRQ(RTMR_IRQn);
```
Comment (andyross) [inline drivers/timer/mchp_xec_rtos_timer.c]: Why hit the NVIC?  That bit should clear automatically once the CPU interrupt has been delivered and the source device is deasserted, right?  In general ARM drivers don't need to do this.

File: ext/hal/microchip/mec/mec1501/component/timer.h [hunk 3]
```diff
@@ -216,24 +264,66 @@ typedef struct htmr_regs
...
-	__IOM uint32_t COMP1;	/*!< (@ 0x0000002C) CCT Compare 1	  */
+	__IOM uint32_t FREE_RUN;	/*!< (@ 0x0000000C) CCT Free run counter */
+	__IOM uint32_t CAP0;		/*!< (@ 0x00000010) CCT Capture 0 */
+	__IOM uint32_t CAP1;		/*!< (@ 0x00000014) CCT Capture 1 */
+	__IOM uint32_t CAP2;		/*!< (@ 0x00000018) CCT Capture 2 */
+	__IOM uint32_t CAP3;		/*!< (@ 0x0000001C) CCT Capture 3 */
+	__IOM uint32_t CAP4;		/*!< (@ 0x00000020) CCT Capture 4 */
+	__IOM uint32_t CAP5;		/*!< (@ 0x00000024) CCT Capture 5 */
+	__IOM uint32_t COMP0;		/*!< (@ 0x00000028) CCT Compare 0 */
+	__IOM uint32_t COMP1;		/*!< (@ 0x0000002C) CCT Compare 1 */
```
Comment (andyross) [inline ext/hal/microchip/mec/mec1501/component/timer.h]: There's a ton of whitespace-only change going on here.  Would be cleaner to leave the space conventions unchanged.

File: drivers/timer/mchp_xec_rtos_timer.c [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+	mchp_pcr_periph_reset(PCR_RTMR);
+	timer_clear_intr();
+
+	last_load = CYC_PER_TICK;
+	cycle_count = 0U;
+	announced_cycles = 0U;
+	elapsed_counts = 0U;
+
+	IRQ_CONNECT(RTMR_IRQn, 1, xec_rtos_timer_isr, 0, 0);
+	irq_enable(RTMR_IRQn);
```
Comment (andyross) [inline drivers/timer/mchp_xec_rtos_timer.c]: Might consider moving irq_enable() to the very end of the process to avoid receiving spurious interrupts from a device in unknown state.

File: drivers/timer/mchp_xec_rtos_timer.c [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+/* Minimum cycles in the future to try to program.  Note that this is
+ * NOT simply "enough cycles to get the counter read and reprogrammed
+ * reliably" -- it becomes the minimum value of the LOAD register, and
+ * thus reflects how much time we can reliably see expire between
+ * calls to elapsed() to read the COUNTFLAG bit.  So it needs to be
+ * set to be larger than the maximum time the interrupt might be
+ * masked.  Choosing a fraction of a tick is probably a good enough
+ * default, with an absolute minimum of 1k cyc.
+ */
+#define MIN_DELAY MAX(32, (CYC_PER_TICK/16))
```
Comment (andyross) [inline drivers/timer/mchp_xec_rtos_timer.c]: Hey, I recognize that comment.  I wrote all that for SysTick, and while the architecture here seems similar, you should at least update the comment to reflect the hardware actually in use. :)  You don't have a COUNTFLAG bit, the comment still talks about a 1024 cycle minimum, etc....

Also note that while this shortcut was probably OK for SysTick running at dozens of MHz, on a much slower counter you're losing precision in those 32 cycles that you'd ideally like to keep (c.f. the carefully crafted cycle-optimal change just submitted for nrf5).  Ideally you should be able to read/check/set/check the counter to lose at most one cycle.

Though I do agree that this is harder than a free running counter to get right.

File: drivers/timer/mchp_xec_rtos_timer.c [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+static ALWAYS_INLINE void timer_counter_set(u32_t val)
+{
+	RTMR_REGS->CNT = val & COUNTER_MAX;
+}
+
+static ALWAYS_INLINE void timer_start(void)
+{
+	RTMR_REGS->CTRL = MCHP_RTMR_CTRL_BLK_EN + MCHP_RTMR_CTRL_HW_HALT_EN
+			  + MCHP_RTMR_CTRL_START;
```
Comment (andyross) [inline drivers/timer/mchp_xec_rtos_timer.c]: "+" seems like a weird operation for what certainly seems like a bitmask.  Prefer "|" to conform with standard idioms.

File: drivers/timer/mchp_xec_rtos_timer.c [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+static u32_t elapsed(void)
+{
+	u32_t tdiff = (last_load - (RTMR_REGS->CNT & COUNTER_MAX));
+	u32_t elapsed = 0;
+
+	if (tdiff != elapsed_counts) {
+		elapsed_counts = tdiff;
+		elapsed = tdiff;
+	}
+	return elapsed;
```
Comment (andyross) [inline drivers/timer/mchp_xec_rtos_timer.c]: This implementation looks wrong to me.  The return value of z_clock_elapsed()  (for which this is the basic engine) is the number of ticks that have elapsed since the last call to z_clock_announce().

This will do the right thing once.  Then it will start returning zeros (!) until the next cycle ticks over, at which point it will return the correct value.  Once.  Then it will break until the next cycle, etc...

If you need those semantics elsewhere in the code, you may need to split out two different elapsed() variants.

File: drivers/timer/mchp_xec_rtos_timer.c [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+void xec_rtos_timer_isr(void *arg)
+{
+	ARG_UNUSED(arg);
+	u32_t dticks;
+
+#ifdef DEBUG_RTOS_TIMER
+	GPIO_CTRL_REGS->CTRL_0015 ^= (1ul << 16);
+#endif
```
Comment (andyross) [inline drivers/timer/mchp_xec_rtos_timer.c]: Probably not a good idea to seize an unrelated GPIO in this driver without documentation.  I'd remove this part before merge.

File: drivers/timer/mchp_xec_rtos_timer.c [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+int z_clock_driver_init(struct device *device)
+{
+	mchp_pcr_periph_slp_ctrl(PCR_RTMR, 0);
+	mchp_pcr_periph_reset(PCR_RTMR);
+	timer_clear_intr();
+
+	last_load = CYC_PER_TICK;
+	cycle_count = 0U;
+	announced_cycles = 0U;
+	elapsed_counts = 0U;
```
Comment (andyross) [inline drivers/timer/mchp_xec_rtos_timer.c]: Those are all in .bss.  You can rely on them being zero already.  Remove.

File: drivers/timer/mchp_xec_rtos_timer.c [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+	delay = ((delay + CYC_PER_TICK - 1) / CYC_PER_TICK) * CYC_PER_TICK;
+	last_load = delay - (cycle_count - announced_cycles);
+
+	timer_stop();
+	delay = last_load;
+	timer_clear_intr();
+	timer_preload_set(delay);
+	timer_counter_set(delay);
+	timer_start();
+
```
Comment (andyross) [inline drivers/timer/mchp_xec_rtos_timer.c]: Does the hardware really require you do this dance every time you want to change the timeout value?  That's... sort of unfortunate.   This call gets made often, sometimes with tight latency requirements, and we really don't want it halting the counter.

Think about it: one of the worst side effects imaginable of code that spins managing a zillion small timeouts is that the underlying clock starts to slow down!

Other architectures all allow the counter to be set while it's ticking.  My hope is this one does too.

**Review by andyross — CHANGES_REQUESTED** [review]
Notes just on the timer driver.   There's one spot with elapsed() that I'm pretty sure is wrong.  Everything else is just nitpicking.

File: drivers/timer/mchp_xec_rtos_timer.c (hunk 10 above)
Comment (scottwcpg) [reply drivers/timer/mchp_xec_rtos_timer.c]: Yes, the HW design requires this.  When I didn't do it the timer would not take the new preload or count register values and wrap from 0.  I have confirmed with the designers that there is a race condition that requires disable timer, load count and/or preload, enable timer.

**Comment by scottwcpg** [comment]
General comment. I would prefer to discard this driver and solve the deep sleep timer issue by:
1. Use a high speed timer for all modes except deep sleep.
2. For deep sleep. Shut down the high speed timer and start a 32KHz based timer for deep sleep wake timeout.
3. On wake shut off 32K timer and adjust high speed timer count for length of time system was asleep.

File: drivers/timer/mchp_xec_rtos_timer.c (hunk 10 above)
Comment (andyross) [reply drivers/timer/mchp_xec_rtos_timer.c]: Yeah, then that definitely deserves a comment to prevent people like me from coming along and trying to "fix" it. :)

**Comment by andyross** [comment]
Regarding using the device only for idle: there's an architecture for that in subsys/power/power.c, see _sys_suspend().  It gets a "ticks" argument from the kernel idle loop indicating how long it should wait for the next timeout.

This code isn't well exercised though, so you may have some plumbing work to do.

**Comment by scottwcpg** [comment]
Hi Everyone,
Thanks for the feed back...
I'm abandoning this PR. Why? I can't make it work even after 3 weeks.  I guess I can't understand what the timer driver is supposed to do.  The only timer drive I can get to work is a high speed free run counter with a compare register because the HW is similar to the other drivers.

File: drivers/timer/mchp_xec_rtos_timer.c [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+#if 0
+#define CYC_PER_TICK (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC	\
+		      / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+#else
+
+#define CYC_PER_TICK_M ((CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC * 256)	\
+		      / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+#if (CYC_PER_TICK_M & 0xFF) > 0x7F
+#define CYC_PER_TICK ((CYC_PER_TICK_M >> 8) + 1)
```
Comment (albertofloyd) [inline drivers/timer/mchp_xec_rtos_timer.c]: @scottwcpg This additional definition seems to introduce some deviation. See results from experiment
These are average values measured on logic analyzer when toggling a GPIO

ksleep(ms)	CYC_PER_TICK=33  CYC_PER_TICK=32
200		203.1			196.8
150		152.7			148.1
100		102.5			99.42
50		52.31			50.73
25		27.2			26.39
10		12.14			11.78
5		7.121			6.907
1		3.104			3.02

**Comment by scottwcpg** [comment]
Hi,
I re-implemented the driver based upon Andy's recommendation to use the new local APIC driver as an example. In addition, I performed a number of experiments with the hardware and it does not like to be reprogrammed while its is counting.  One issue was the ISR re-starts the timer and then announces ticks to the kernel.  The announce API will call into the driver to set a new timeout.  I believe if this happens before the first 32KHz count has elapsed from timer restart then the hardware will sometimes take up to an additional 30.5 us to synchronize and start counting again.  The drive is tuned to be more accurate for small sleep requests: 1, 5, 10ms.  For > 50ms the application will need to adjust and ask for 1 to 2 extra ticks.

File: drivers/timer/mchp_xec_rtos_timer.c [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+	RTMR_REGS->CTRL = MCHP_RTMR_CTRL_BLK_EN;
+	RTMR_REGS->CNT;
+	RTMR_REGS->CTRL = 0;
+
+	GIRQ23_REGS->SRC = MCHP_RTMR_GIRQ_VAL;
+	NVIC_ClearPendingIRQ(RTMR_IRQn);
+
+	RTMR_REGS->PRLD = cached_icr;
+
+	IRQ_CONNECT(RTMR_IRQn, 1, xec_rtos_timer_isr, 0, 0);
```
Comment (andrewboie) [inline drivers/timer/mchp_xec_rtos_timer.c]: This is causing a lot of build errors:

```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/mchp_xec_rtos_timer.c:287:28: error: 'xec_rtos_timer_isr' undeclared (first use in this function)
```

board is mec15xxevb_assy6853

Comment (franciscomunoz) [reply drivers/timer/mchp_xec_rtos_timer.c]: @scottwcpg  FYI

TICKLESS_KERNEL(=n) "Tickless kernel"

fails like Andrew said above

[ 34%] Building C object zephyr/CMakeFiles/zephyr.dir/drivers/timer/mchp_xec_rtos_timer.c.obj
/home/francisco/mchip_zephyr/zephyr/drivers/timer/mchp_xec_rtos_timer.c: In function 'z_clock_driver_init':
/home/francisco/mchip_zephyr/zephyr/drivers/timer/mchp_xec_rtos_timer.c:256:151: error: 'xec_rtos_timer_isr' undeclared (first use in this function)
  IRQ_CONNECT(RTMR_IRQn, 1, xec_rtos_timer_isr, 0, 0);

File: soc/arm/microchip_mec/mec1501/soc.c [hunk 13]
```diff
@@ -136,6 +136,11 @@ static int soc_init(struct device *dev)
 	isave = __get_PRIMASK();
 	__disable_irq();
 
+	/* Hack: disable Cortex-M Serial Wire Viewer, only use
+	 * the two SWD pin (TMS & TCLK)
+	 */
+	ECS_REGS->DEBUG_CTRL = 0x05;
```
Comment (franciscomunoz) [inline soc/arm/microchip_mec/mec1501/soc.c]: We currently have this in the boards/arm boards/arm/mec15xxevb_assy6853/pinmux.c . We put it there to emphasize that we are releasing TDI/TDO. If you want you can set 0x05 there.

File: soc/arm/microchip_mec/mec1501/power.c [hunk 14]
```diff
@@ -0,0 +1,228 @@
+/*
```
Comment (franciscomunoz) [inline soc/arm/microchip_mec/mec1501/power.c]: Remove this file and use it in another PR related to power management.

File: soc/arm/microchip_mec/mec1501/CMakeLists.txt [hunk 15]
```diff
@@ -7,3 +7,7 @@
 zephyr_sources(
   soc.c
   )
+
+zephyr_sources_ifdef(CONFIG_SYS_POWER_MANAGEMENT
```
Comment (franciscomunoz) [inline soc/arm/microchip_mec/mec1501/CMakeLists.txt]: Undo this change and place it in another PR

File: dts/bindings/timer/microchip,xec-rtos-timer.yaml [hunk 16]
```diff
@@ -0,0 +1,37 @@
+#
+# Copyright (c) 2019 Microchip Technology Inc.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+---
+title: Microchip RTOS timer
```
Comment (franciscomunoz) [inline dts/bindings/timer/microchip,xec-rtos-timer.yaml]: Change to : Microchip XEC RTOS timer

File: dts/bindings/timer/microchip,xec-rtos-timer.yaml [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+version: 0.1
+
+description: >
+    This is a representation of the Microchip RTOS timer node
```
Comment (franciscomunoz) [inline dts/bindings/timer/microchip,xec-rtos-timer.yaml]: Change the sentence to : This is a representation of the Microchip XEC RTOS timer node
just to make the sentence more specific

File: boards/arm/mec15xxevb_assy6853/Kconfig.defconfig [hunk 18]
```diff
@@ -72,4 +72,32 @@ config I2C_XEC_2
...
+# If the kernel has timer support, enable both clock control and timer
+config CLOCK_CONTROL
+	default y
+
+config MCHP_XEC_RTOS_TIMER
+	default y
+
+endif # SYS_CLOCK_EXISTS
+
+config SYS_POWER_MANAGEMENT
```
Comment (franciscomunoz) [inline boards/arm/mec15xxevb_assy6853/Kconfig.defconfig]: Remove the Kconfig flags and code changes related to power management, and create a separate PR to investigate power management.

**Review by franciscomunoz — CHANGES_REQUESTED** [review]

**Comment by scottwcpg** [comment]
Hi,
I made the changes Francisco requested, squashed the two commits into one, and pushed it.

File: boards/arm/mec15xxevb_assy6853/Kconfig.defconfig (hunk 18 above)
Comment (scottwcpg) [reply boards/arm/mec15xxevb_assy6853/Kconfig.defconfig]: Done

File: dts/bindings/timer/microchip,xec-rtos-timer.yaml (hunk 16 above)
Comment (scottwcpg) [reply dts/bindings/timer/microchip,xec-rtos-timer.yaml]: Done

File: soc/arm/microchip_mec/mec1501/soc.c (hunk 13 above)
Comment (scottwcpg) [reply soc/arm/microchip_mec/mec1501/soc.c]: Done

File: drivers/timer/mchp_xec_rtos_timer.c (hunk 12 above)
Comment (scottwcpg) [reply drivers/timer/mchp_xec_rtos_timer.c]: Done. Checked build with tickless disabled is OK.

**Review by franciscomunoz — COMMENTED** [review]
Looks fine to me. Let's just continue testing a few scenarios more.

**Review by franciscomunoz — APPROVED** [review]

File: drivers/timer/mchp_xec_rtos_timer.c [hunk 19]
```diff
@@ -0,0 +1,297 @@
...
+		full_ticks = n - 1;
+	}
+
+	/*
+	 * RTMR frequency is fixed at 32KHz resulting in truncation errors.
+	 * Tune the denominator to take into account delay in caller and this
+	 * routine.
+	 */
+	full_cycles = (full_ticks * CYCLES_PER_TICK)
+		      + ((full_ticks * CPT_FRACT) / 1024UL);
```
Comment (andyross) [inline drivers/timer/mchp_xec_rtos_timer.c]: More work to try to handle precision loss that goes away with a proper default.

File: drivers/timer/mchp_xec_rtos_timer.c [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+	u32_t ccr;
+	u32_t ticks;
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	ccr = RTMR_REGS->CNT;
+	ticks = total_cycles - last_announcement;
+	ticks += cached_icr - ccr;
+	k_spin_unlock(&lock, key);
+	ticks /= CYCLES_PER_TICK;
```
Comment (andyross) [inline drivers/timer/mchp_xec_rtos_timer.c]: Style: you're reading a value in cycles and assigning it to a variable named "ticks", which you then convert at the end of the function.  Would be clearer if you named the variable "cyc" and then did a "`return cyc / CYCLES_PER_TICK;`"

**Review by andyross — APPROVED** [review]
Some nitpicking about units/precision/defaults, and style notes.  Nothing looks wrong to me.

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by andrewboie** [comment]
@scottwcpg this is failing CI:

```
FAILED: ccache /opt/sdk/zephyr-sdk-0.10.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DBUILD_VERSION=zephyr-v1.14.0-2523-gaab085bab2e1 -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I../../../../../../../kernel/include -I../../../../../../../arch/arm/include -I../../../../../../../include -I../../../../../../../include/drivers -Izephyr/include/generated -I../../../../../../../soc/arm/microchip_mec/mec1501 -I../../../../../../../lib/libc/minimal/include -I../../../../../../../ext/hal/cmsis/Include -I../../../../../../../ext/hal/microchip/mec/common -I../../../../../../../ext/hal/microchip/mec/mec1501 -isystem /opt/sdk/zephyr-sdk-0.10.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/include -isystem /opt/sdk/zephyr-sdk-0.10.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/include-fixed -Os -nostdinc -imacros/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/out-3nd-pass/mec15xxevb_assy6853/samples/subsys/console/echo/sample.subsys.console.echo/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -mthumb -mcpu=cortex-m4 -imacros/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wno-unused-but-set-variable -Werror=implicit-int -Werror -Wno-deprecated-declarations -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/samples/subsys/console/echo=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr=ZEPHYR_BASE -ffunction-sections -fdata-sections -mabi=aapcs -march=armv7e-m -std=c99 -MD -MT zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_ns16550.c.obj -MF zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_ns16550.c.obj.d -o zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_ns16550.c.obj   -c /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/serial/uart_ns16550.c
In file included from ../../../../../../../include/arch/arm/arch.h:27,
                 from ../../../../../../../include/arch/cpu.h:17,
                 from ../../../../../../../include/kernel_includes.h:34,
                 from ../../../../../../../include/kernel.h:17,
                 from /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/serial/uart_ns16550.c:24:
zephyr/include/generated/uart_ns16550_port_2.h: In function 'irq_config_func_2':
zephyr/include/generated/generated_dts_board_fixups.h:53:41: error: 'DT_INST_2_NS16550_IRQ_0_PRIORITY' undeclared (first use in this function); did you mean 'DT_INST_0_NS16550_IRQ_0_PRIORITY'?
 #define DT_UART_NS16550_PORT_2_IRQ_PRI  DT_INST_2_NS16550_IRQ_0_PRIORITY
                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../../../../../include/arch/arm/cortex_m/irq.h:85:28: note: in definition of macro 'Z_ARCH_IRQ_CONNECT'
  z_irq_priority_set(irq_p, priority_p, flags_p); \
                            ^~~~~~~~~~
zephyr/include/generated/uart_ns16550_port_2.h:94:2: note: in expansion of macro 'IRQ_CONNECT'
  IRQ_CONNECT(DT_UART_NS16550_PORT_2_IRQ,
  ^~~~~~~~~~~
zephyr/include/generated/uart_ns16550_port_2.h:95:7: note: in expansion of macro 'DT_UART_NS16550_PORT_2_IRQ_PRI'
       DT_UART_NS16550_PORT_2_IRQ_PRI,
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
zephyr/include/generated/generated_dts_board_fixups.h:53:41: note: each undeclared identifier is reported only once for each function it appears in
 #define DT_UART_NS16550_PORT_2_IRQ_PRI  DT_INST_2_NS16550_IRQ_0_PRIORITY
                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../../../../../include/arch/arm/cortex_m/irq.h:85:28: note: in definition of macro 'Z_ARCH_IRQ_CONNECT'
  z_irq_priority_set(irq_p, priority_p, flags_p); \
                            ^~~~~~~~~~
zephyr/include/generated/uart_ns16550_port_2.h:94:2: note: in expansion of macro 'IRQ_CONNECT'
  IRQ_CONNECT(DT_UART_NS16550_PORT_2_IRQ,
  ^~~~~~~~~~~
zephyr/include/generated/uart_ns16550_port_2.h:95:7: note: in expansion of macro 'DT_UART_NS16550_PORT_2_IRQ_PRI'
       DT_UART_NS16550_PORT_2_IRQ_PRI,
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
```


```
FAILED: ccache /opt/sdk/zephyr-sdk-0.10.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DBUILD_VERSION=zephyr-v1.14.0-2523-gaab085bab2e1 -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -I../../../../../../kernel/include -I../../../../../../arch/arm/include -I../../../../../../include -I../../../../../../include/drivers -Izephyr/include/generated -I../../../../../../soc/arm/microchip_mec/mec1501 -I../../../../../../lib/libc/minimal/include -I../../../../../../ext/hal/cmsis/Include -I../../../../../../ext/hal/microchip/mec/common -I../../../../../../ext/hal/microchip/mec/mec1501 -I../../../../../../subsys/testsuite/include -isystem /opt/sdk/zephyr-sdk-0.10.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/include -isystem /opt/sdk/zephyr-sdk-0.10.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/include-fixed -Os -nostdinc -imacros/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/out-3nd-pass/mec15xxevb_assy6853/tests/benchmarks/sys_kernel/benchmark.kernel/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -mthumb -mcpu=cortex-m4 -imacros/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wno-unused-but-set-variable -Werror=implicit-int -Werror -Wno-deprecated-declarations -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/benchmarks/sys_kernel=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr=ZEPHYR_BASE -ffunction-sections -fdata-sections -mabi=aapcs -march=armv7e-m -std=c99 -MD -MT zephyr/CMakeFiles/zephyr.dir/drivers/timer/mchp_xec_rtos_timer.c.obj -MF zephyr/CMakeFiles/zephyr.dir/drivers/timer/mchp_xec_rtos_timer.c.obj.d -o zephyr/CMakeFiles/zephyr.dir/drivers/timer/mchp_xec_rtos_timer.c.obj   -c /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/mchp_xec_rtos_timer.c
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/mchp_xec_rtos_timer.c: In function 'xec_rtos_timer_isr':
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/mchp_xec_rtos_timer.c:234:2: error: implicit declaration of function 'timer_restart'; did you mean 'k_timer_start'? [-Werror=implicit-function-declaration]
  timer_restart(cached_icr);
  ^~~~~~~~~~~~~
  k_timer_start
cc1: all warnings being treated as errors
ninja: build stopped: subcommand failed.
```

**Comment by scottwcpg** [comment]
Fixed build when TICKLESS is disabled.

**Comment by andrewboie** [comment]
```
FAILED: ccache /opt/sdk/zephyr-sdk-0.10.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DBUILD_VERSION=zephyr-v1.14.0-2583-g6361ddbb5d81 -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -I../../../../../../kernel/include -I../../../../../../arch/arm/include -I../../../../../../include -I../../../../../../include/drivers -Izephyr/include/generated -I../../../../../../soc/arm/microchip_mec/mec1501 -I../../../../../../lib/libc/minimal/include -I../../../../../../ext/hal/cmsis/Include -I../../../../../../ext/hal/microchip/mec/common -I../../../../../../ext/hal/microchip/mec/mec1501 -I../../../../../../subsys/testsuite/include -I../../../../../../subsys/testsuite/ztest/include -isystem /opt/sdk/zephyr-sdk-0.10.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/include -isystem /opt/sdk/zephyr-sdk-0.10.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/include-fixed -Os -nostdinc -imacros/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/out-3nd-pass/mec15xxevb_assy6853/tests/kernel/fatal/kernel.common.stack_sentinel/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -mthumb -mcpu=cortex-m4 -imacros/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wno-unused-but-set-variable -Werror=implicit-int -Werror -Wno-deprecated-declarations -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/kernel/fatal=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr=ZEPHYR_BASE -ffunction-sections -fdata-sections -mabi=aapcs -march=armv7e-m -std=c99 -MD -MT zephyr/CMakeFiles/zephyr.dir/drivers/timer/mchp_xec_rtos_timer.c.obj -MF zephyr/CMakeFiles/zephyr.dir/drivers/timer/mchp_xec_rtos_timer.c.obj.d -o zephyr/CMakeFiles/zephyr.dir/drivers/timer/mchp_xec_rtos_timer.c.obj   -c /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/mchp_xec_rtos_timer.c
In file included from ../../../../../../include/toolchain.h:39,
                 from ../../../../../../include/kernel_includes.h:19,
                 from ../../../../../../include/kernel.h:17,
                 from ../../../../../../include/device.h:11,
                 from ../../../../../../include/drivers/timer/system_timer.h:24,
                 from /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/mchp_xec_rtos_timer.c:8:
../../../../../../include/toolchain/gcc.h:28:37: error: static assertion failed: "MCHP XEC RTOS timer: Non-tickless bad CYCLES_PER_TICK"
 #define BUILD_ASSERT_MSG(EXPR, MSG) _Static_assert(EXPR, MSG)
                                     ^~~~~~~~~~~~~~
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/mchp_xec_rtos_timer.c:218:1: note: in expansion of macro 'BUILD_ASSERT_MSG'
 BUILD_ASSERT_MSG(CYCLES_PER_TICK >= 2,
 ^~~~~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
```

**Comment by andrewboie** [comment]
The issue with DT_INST_2_NS16550_IRQ_0_PRIORITY is broken in master, I opened #17751

**Comment by andrewboie** [comment]
DT_INST_2_NS16550_IRQ_0_PRIORITY issue fixed.
@scottwcpg please rebase.
